### PR TITLE
Wait for handover state update before serving the requests

### PR DESCRIPTION
**Pull request by yux0**
## What changed?
Wait for handover state update before serving the requests.

## Why?
We need to wait on handover state update before the redirect interceptor. It will serve the request once the replication state is updated and routes to the correct endpoint.

## How did you test it?
TODO

### CODE REVIEW

File: common/rpc/interceptor/namespace_validator.go [hunk 1]
```diff
@@ -237,10 +242,35 @@ func (ni *NamespaceValidatorInterceptor) StateValidationIntercept(
...
+	handler grpc.UnaryHandler,
+	nsEntry *namespace.Namespace,
+) (interface{}, error) {
+
+	var response interface{}
+	op := func(ctx context.Context) error {
+		err := ni.checkReplicationState(nsEntry, info.FullMethod)
+		if err != nil {
+			var nsErr error
+			nsEntry, nsErr = ni.namespaceRegistry.RefreshNamespaceById(nsEntry.ID())
```
Comment (yycptt) [inline common/rpc/interceptor/namespace_validator.go:261]: hmmm this will keep refreshing the namespace? I mean all requests will perform a refresh upon retry.

also if the handover error is from history service then the refresh here is unnecessary

File: common/rpc/interceptor/namespace_validator.go [hunk 2]
```diff
@@ -237,10 +242,35 @@ func (ni *NamespaceValidatorInterceptor) StateValidationIntercept(
...
+			}
+			return err
+		}
+
+		response, err = handler(ctx, req)
 		return err
 	}
-	return ni.checkReplicationState(namespaceEntry, fullMethod)
+	// Only retry on ErrNamespaceHandover, other errors will be handled by the retry interceptor
+	err := backoff.ThrottleRetryContext(ctx, op, handoverRetryPolicy, common.IsNamespaceHandoverError)
```
Comment (yycptt) [inline common/rpc/interceptor/namespace_validator.go:272]: I also worry this will cause server api latency alert to fire during the handover, we should probably exclude this backoff from nouserlatency metric.

File: common/rpc/interceptor/namespace_validator.go [hunk 3]
```diff
@@ -237,10 +242,35 @@ func (ni *NamespaceValidatorInterceptor) StateValidationIntercept(
...
+				return nsErr
+			}
+			return err
+		}
+
+		response, err = handler(ctx, req)
 		return err
 	}
-	return ni.checkReplicationState(namespaceEntry, fullMethod)
+	// Only retry on ErrNamespaceHandover, other errors will be handled by the retry interceptor
```
Comment (yycptt) [inline common/rpc/interceptor/namespace_validator.go:271]: shall we just let retry interceptor handle the error? and this interceptor just block until ns is no longer in handover state?

Comment (yux0) [reply common/rpc/interceptor/namespace_validator.go:271]: I did not go with this because retry interceptor should be the most inner interceptor. If that is the case, can I add a handover interceptor after the retry interceptor?

File: common/rpc/interceptor/namespace_validator.go (hunk 1 above)
Comment (yux0) [reply common/rpc/interceptor/namespace_validator.go:261]: Make sense. We can also just wait for ns refresh not do this proactive refresh.

File: common/rpc/interceptor/namespace_validator.go (hunk 2 above)
Comment (yux0) [reply common/rpc/interceptor/namespace_validator.go:272]: ack

File: common/util.go [hunk 4]
```diff
@@ -357,6 +357,10 @@ func IsServiceHandlerRetryableError(err error) bool {
 	return false
 }
 
+func IsNamespaceHandoverError(err error) bool {
+	return err.Error() == ErrNamespaceHandover.Error()
```
Comment (yycptt) [inline common/util.go:361]: nit: can you replace other places that perform the same change to use this func? e.g. L341 above.

File: common/rpc/interceptor/telemetry.go [hunk 5]
```diff
@@ -215,10 +215,14 @@ func (ti *TelemetryInterceptor) RecordLatencyMetrics(ctx context.Context, startT
 		userLatencyDuration = time.Duration(val)
 		metrics.ServiceLatencyUserLatency.With(metricsHandler).Record(userLatencyDuration)
 	}
+	handoverRetryLatency := time.Duration(0)
+	if val, ok := metrics.ContextCounterGet(ctx, metrics.NamespaceHandoverRetryLatency.Name()); ok {
+		handoverRetryLatency = time.Duration(val)
```
Comment (yycptt) [inline common/rpc/interceptor/telemetry.go:220]: so we don't actually emit metric for this latency?

File: common/rpc/interceptor/namespace_validator.go [hunk 6]
```diff
@@ -237,10 +243,41 @@ func (ni *NamespaceValidatorInterceptor) StateValidationIntercept(
...
+func (ni *NamespaceValidatorInterceptor) handleRequestWithHandoverRetry(
+	ctx context.Context,
+	req interface{},
+	info *grpc.UnaryServerInfo,
+	handler grpc.UnaryHandler,
+	nsEntry *namespace.Namespace,
+) (interface{}, error) {
+
+	var response interface{}
+	var retryCount int
```
Comment (yycptt) [inline common/rpc/interceptor/namespace_validator.go:258]:
```suggestion
	var attempt int
```

File: common/rpc/interceptor/namespace_validator.go [hunk 7]
```diff
@@ -237,10 +243,41 @@ func (ni *NamespaceValidatorInterceptor) StateValidationIntercept(
...
+		err := ni.checkReplicationState(nsEntry, info.FullMethod)
+		if err != nil {
+			return err
+		}
+
+		response, err = handler(ctx, req)
+		if retryCount > 1 {
+			retryLatency = retryLatency + handoverRetryPolicy.ComputeNextDelay(0, retryCount, nil)
+		}
+		retryCount++
```
Comment (yycptt) [inline common/rpc/interceptor/namespace_validator.go:270]: it's a bit hard to follow as it's calculating the retryLatency before the current attempt, but the code lives after the handler call...

File: common/rpc/interceptor/namespace_validator.go (hunk 3 above)
Comment (yycptt) [reply common/rpc/interceptor/namespace_validator.go:273]: I couldn't remember why we moved retry interceptor to be the most inner interceptor, but I think there must be a reason. I think we should fully understand that to make sure the retry logic here won't cause any issue.

> can I add a handover interceptor after the retry interceptor?

hmmm then why not just retry handover error in retry interceptor?

I guess I was proposing something different in my comment, i.e. block on ns state change instead of retry (by registering a callback to namespace registry).

File: common/rpc/interceptor/telemetry.go (hunk 5 above)
Comment (yux0) [reply common/rpc/interceptor/telemetry.go:220]: Follow the command to exclude the backoff retry delay from the no user latency.

Comment (yux0) [reply common/rpc/interceptor/telemetry.go:220]: We need to chat about this.

File: common/rpc/interceptor/namespace_validator.go (hunk 3 above)
Comment (yux0) [reply common/rpc/interceptor/namespace_validator.go:273]: I cannot wait for ns callback as the error might return from history nodes and there is no callback from frontend.

File: common/rpc/interceptor/namespace_handover.go [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+
+	startTime := time.Now()
+	namespaceData, err := i.namespaceRegistry.GetNamespace(namespaceName)
+	if err != nil {
+		return 0, err
+	}
+	if namespaceData.ReplicationState() == enumspb.REPLICATION_STATE_HANDOVER {
+		cbID := uuid.New()
+		waitReplicationStateUpdate := make(chan struct{})
+		i.namespaceRegistry.RegisterStateChangeCallback(cbID, func(ns *namespace.Namespace, deletedFromDb bool) {
```
Comment (yycptt) [inline common/rpc/interceptor/namespace_handover.go:104]: we probably should be change the impl to listen on changes for a specific namespace...
Can be done in a later PR.

File: common/rpc/interceptor/namespace_handover.go [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+			if ns.State() != enumspb.NAMESPACE_STATE_REGISTERED ||
+				deletedFromDb ||
+				ns.ReplicationState() != enumspb.REPLICATION_STATE_HANDOVER ||
+				!ns.IsGlobalNamespace() {
+				// Stop wait on state change if:
+				// 1. namespace is deleting/deleted
+				// 2. namespace is not in handover
+				// 3. namespace is not global
+				select {
+				case <-waitReplicationStateUpdate:
```
Comment (yycptt) [inline common/rpc/interceptor/namespace_handover.go:117]: what's the purpose of this line? I don't see other senders on the channel.

File: common/rpc/interceptor/namespace_handover.go [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+	methodName := api.MethodName(info.FullMethod)
+	namespaceName, err := GetNamespaceName(i.namespaceRegistry, req)
+	if err != nil {
+		return nil, err
+	}
+
+	if i.enabledForNS(namespaceName.String()) {
+		startTime := i.timeSource.Now()
+		defer func() {
+			metrics.HandoverWaitLatency.With(i.metricsHandler).Record(time.Since(startTime))
```
Comment (yycptt) [inline common/rpc/interceptor/namespace_handover.go:76]: I guess we only want to emit the metric when namespace actually waited on the handover state.

File: service/frontend/fx.go [hunk 11]
```diff
@@ -267,6 +269,7 @@ func GrpcServerOptionsProvider(
 		namespaceLogInterceptor.Intercept, // TODO: Deprecate this with a outer custom interceptor
 		metrics.NewServerMetricsContextInjectorInterceptor(),
 		authInterceptor.Intercept,
+		namespaceHandoverInterceptor.Intercept,
```
Comment (yycptt) [inline service/frontend/fx.go:272]: Can you add some comments explaining why it must be in this place?

Comment (yycptt) [reply service/frontend/fx.go:272]: Is the handover state checking logic in StateValidationIntercept removed?

File: service/frontend/fx.go [hunk 12]
```diff
@@ -267,6 +269,7 @@ func GrpcServerOptionsProvider(
...
 		redirectionInterceptor.Intercept,
```
Comment (yycptt) [inline service/frontend/fx.go:273]: I guess theoretically we can retry in redirection Interceptor? since it has the knowledge of whether redirect will actually happen or not?

File: common/rpc/interceptor/namespace_handover.go [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+				childCancelFn()
+			}()
+		}
+		select {
+		case <-childCtx.Done():
+			err = common.ErrNamespaceHandover
+		case <-waitReplicationStateUpdate:
+		}
+		i.namespaceRegistry.UnregisterStateChangeCallback(cbID)
+		if err != nil {
```
Comment (yycptt) [inline common/rpc/interceptor/namespace_handover.go:138]: nit: we don't need this check?

File: common/rpc/interceptor/namespace_handover.go (hunk 9 above)
Comment (yux0) [reply common/rpc/interceptor/namespace_handover.go:117]: in case the channel is already closed.

File: service/frontend/fx.go (hunk 12 above)
Comment (yux0) [reply service/frontend/fx.go:273]: I guess so but the logic could be weird. What should we do if the redirect will not happen?

File: service/frontend/fx.go (hunk 11 above)
Comment (yux0) [reply service/frontend/fx.go:272]: I did not remove in this PR as I added the feature flag. Once we can confident we can remove it.

Comment (yycptt) [reply service/frontend/fx.go:274]: do you want to disabled that the check in StateValidationIntercept if the feature flag is enabled?

File: common/rpc/interceptor/namespace_handover.go [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+	return handler(ctx, req)
+}
+
+func (i *NamespaceHandoverInterceptor) waitNamespaceHandoverUpdate(
+	ctx context.Context,
+	namespaceName namespace.Name,
+	methodName string,
+) (waitTime *time.Duration, retErr error) {
+	if _, ok := allowedMethodsDuringHandover[methodName]; ok {
+		return
```
[the remark this answers is not in the document]
Comment (yux0) [reply common/rpc/interceptor/namespace_handover.go]: done

File: service/frontend/fx.go (hunk 11 above)
Comment (yux0) [reply service/frontend/fx.go:273]: Yes. Once I validate the function and set default to enable I will remove it
